### PR TITLE
fix: cleanup ConversationEventProcessor todos - WPB-10283

### PR DESCRIPTION
**Pull request by netbe**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10283" title="WPB-10283" target="_blank">TaskWPB-10283</a>  Check TODOs and FIXMEs regarding event processing
  </td></table>
  <br />
 

### Issue

This PR cleans up some TODOs in order to avoid passing events to eventProcessor when they can be directly consumed by ConversationEventProcessor.

- Use `ConversationEventProcessor` directly when possible
- Update `processConversationEvents` method from UserSession to save also the context before returning completion
 
---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [ ] Make sure you use the API for UI elements that support large fonts.
- [ ] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [ ] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationRequestStrategy.swift [hunk 1]
```diff
@@ -387,7 +387,10 @@ extension ConversationRequestStrategy: ZMUpstreamTranscoder {
         else {
             return false
         }
-        // TODO: [WPB-10283] [F] check if we need to wait for the processPayload
+
+        // There is one case where you end up here:
+        // 1) selfUser edits the conversation name: a save will be enqueue when user is done editing
+        // Note: when an other user edited the conversation name, ConversationEventProcessor is called directly as EventAsyncConsumer and a save will be done in EventProcessor
```
Comment (samwyndham) [inline wire-ios-request-strategy/Sources/Request Strategies/Conversation/ConversationRequestStrategy.swift]:
```suggestion
        // Note: when another user edited the conversation name, ConversationEventProcessor is called directly as EventAsyncConsumer and a save will be done in EventProcessor
```

File: wire-ios-sync-engine/Source/Data Model/Conversation+MessageDestructionTimeout.swift [hunk 2]
```diff
@@ -57,18 +58,19 @@ extension ZMConversation {
...
 
-        guard let apiVersion = BackendInfo.apiVersion else {
+        guard let apiVersion = BackendInfo.apiVersion,
+                let managedObjectContext else {
             return completion(.failure(WirelessLinkError.unknown))
         }
 
         let request = MessageDestructionTimeoutRequestFactory.set(timeout: Int(timeout.rawValue), for: self, apiVersion: apiVersion)
-        request.add(ZMCompletionHandler(on: managedObjectContext!) { response in
+        request.add(ZMCompletionHandler(on: managedObjectContext) { response in
```
Comment (samwyndham) [inline wire-ios-sync-engine/Source/Data Model/Conversation+MessageDestructionTimeout.swift:67]: I had to look into what magic `ZMCompletionHandler` does!

**Review by samwyndham — APPROVED** [review]
Looks good!

File: wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift [hunk 3]
```diff
@@ -633,9 +633,11 @@ public final class ZMUserSession: NSObject {
...
     // temporary function to simplify call to ConversationEventProcessor
     // might be replaced by something more elegant
-    public func processConversationEvents(_ events: [ZMUpdateEvent]) {
-        WaitingGroupTask(context: self.syncContext) {
-            await self.conversationEventProcessor.processConversationEvents(events)
+    public func processConversationEvents(_ events: [ZMUpdateEvent], completion: (() -> Void)?) {
+        WaitingGroupTask(context: self.syncContext) { [weak self] in
+            guard let self else { return }
+            await self.conversationEventProcessor.processAndSaveConversationEvents(events)
+            completion?()
```
Comment (caldrian) [inline wire-ios-sync-engine/Source/UserSession/ZMUserSession/ZMUserSession.swift]: Shouldn't `completion` be called always?
```suggestion
        WaitingGroupTask(context: self.syncContext) { [weak self] in
            defer { completion?() }
            guard let self else { return }
            await self.conversationEventProcessor.processAndSaveConversationEvents(events)
            completion?()
```
or
```suggestion
        WaitingGroupTask(context: self.syncContext) { [weak self] in
            guard let self else { return completion?() }
            await self.conversationEventProcessor.processAndSaveConversationEvents(events)
            completion?()
```

**Review by caldrian — APPROVED** [review]

**Pull request merged**